### ISSUE TITLE
Add update menu item ... (automate updates)

### ISSUE THREAD

**Issue by bernd-wechner** [sourcegit-scm/sourcegit#426]
I'm really liking SourceGIT and one reason is the vibrant and living nature of the tool, with regular updates and small improvements. Nothing else would encourage me more than that to eventually contribute PRs rather than just post issues ;-). 

One downside of this currently is that it does notify me of a new version and has a button, but it only takes me to the download page. Where I have to find my build and download it and install it each time. Quite a rigmarole.

It's far from uncommon to find software doing that for itself, and just upgrading itself.

Not a trivial task to implement of course, far from it. It needs for the software to know what build it is, and then to choose the appropriate download, download it and finally actually install it which is conceivably a different method for each build and OS type ... and this last one is a part of the challenge as it would conceivably need code contributions from a number of people across all the current supported builds to both implement and test the feature. On some (most) builds it also needs to orchestrate the swap of the running binary - and so that typically needs a separate standalone little utility, that conceivably is copied to a temporary spot and run from there which completes the install before starting the new version.

C'est la vie. That, no doubt, is why we don't have it (yet). But it's a noble vision methinks.

**Comment by love-linger** [comment]
Yes, you are right. Implementing cross-platform automatic updates is indeed a bit difficult. In addition to the things you listed above, there are some of the following:

* On Linux platform, if the software has been included in the official repository, such as: xxxx, the software's self-update function should be blocked. At the same time, since the official repository update may delay (comparing with GitHub Releases), should the existing new version notification function be modified?
* Does the macOS allows us update some of files in the `SourceGit.app`? I have not tested it.

BTW, if you are Windows user, you can install this app from `winget install SourceGit`. Then you can upgrade this app by `winget upgrade SourceGit`

**Comment by bernd-wechner** [comment]
On the Linux front that issue is real but absent on Appimages as a rule, which are gobsmackingly popular because they port across all distros and can manage their own updates ;-).

Thangs for the Windows tip. I am using it on Windows. In one of life's greater ironies, I do my non-commercial development on Linux and have used Gitkraken for many years (it's pretty good) and my commercial work in an office that obligates Windows, and finally made the move from subversion to git a year or two ago, and rolled out SourceTree for use in house. I hated it. SourceTree just bugs me. I can't use Gitkraken because it's a commercial product with a free individual use license and I can't use it on the commercial work so I went shopping in the Foss-boutiques.  The irony being that Gitkraken give us free individual use to so it becomes our "normal" and we come to appreciate its great features so when we do commercial work we'll lobby in house for investment. If only that worked here, we're stuck with a SourceTree recommendation. I shopped around and tried many Foss Git GUI clients and none of them really appealed to me until ... I hit SourceGit. It's clean and slick and easy to use. A total winner.  To find it so actively developed since I started using it, is a plus!

**Comment by ybeapps** [comment]
@bernd-wechner im also new to source-git.
Started from the same reason 
Tried so many git clients until I found this one

BTW, for windows I really like git-extensions. Super powerful tool
But it’s not cross platform
the UI structure of source-git is very similar to git-extensions and works faster

so I started contributing the features I liked most in there :)
And I’m so happy 😀 

@love-linger about auto update, how about creating a branch, and starting to create infra for auto-update.. then contributors can complete for their ENVs or even publish it initially for a single OS and improving over time
I think it will be easy
Step 1: identify current OS and Arch (should be easy) - and also installation kind, for external tools like you mentioned, in such case, a suggestion to run update command will be enough.
Step 2: compare latest version to current (already implemented)
Step 3: download
Step 4: auto install - we’ll help completing it, but we can start with a single OS first

**Comment by aikawayataro** [comment]
The problem is not how to detect and download new releases, but how to handle them on different systems.

I also believe that complexity of detecting something better be avoided. Trying to guess installation type is very tricky, especially on Linux where we have 3 types of releases.
Instead it's better to just bundle this information during build time, there are couple ways to do so but I really liked this one https://stackoverflow.com/a/49537163 - you can define properties via dontet CLI during build, something like `-p:UpdateKind=portable-zip`, the arch and OS comes from `RuntimeIdentifier` and it is currently defined via `-r/--runtime` argument

So I had a little experience in this topic and can say something:

Windows updates are a pain.
You can't remove running executables, but you can rename them. Updates could be done this way, rename all executables including shared libraries (it could be `program.exe` => `program.exe.bak` and so on) and extract the update to the target directory. You still have to clean up old renamed executables, it could be done by killing every instance and spawning delayed deletor process just before it exit itself (or deleting them during application start).

Other way to do update is to create dedicated "updater" program, but what to do when it comes to update "updater" :smile: ?
Thus in most cases the "updater" program is an `.msi` or something installer. Software downloads and run installer, it kills every instance and proceeds to install a new update. I think this is the most sane option for Windows.

Linux AppImage is simple, unlink executable and extract the update.
Installed packages should be updated using disto's package manager and should be packaged by distro maintainer or as a 3rd party repo (#298 like how docker do this for Debian systems).

Couldn't say anything about osx, but if the application gets signed, it can make it to homebrew.

I think I've addressed every point of the auto-update feature, and as you can see, the list of things to choose and consider is big (including stuff for osx I don't know)

If someone will fill the osx gap I would love to try to implement this, but this looks very low-priority to me... The current update process takes only a few extra clicks

**Comment by ybeapps** [comment]
Maybe just pushing the release to some tools will do the job
Each OS has its tools
We don’t need more than that

for MacOS I can help pushing release process to homebrew….

what do you think?

**Comment by love-linger** [comment]
I'm sorry. I'm not very familiar with these... 😂

**Comment by ybeapps** [comment]
Well, it requires a code signed build
So I can help but we need to sign the bundle first

**Comment by ChiahongHong** [comment]
> Maybe just pushing the release to some tools will do the job Each OS has its tools We don’t need more than that
> 
> for MacOS I can help pushing release process to homebrew…. like using this nice GitHub action: https://github.com/marketplace/actions/homebrew-bump-cask requires just creating a simple repo - with predefined structure.. and some ruby script - for installation the GitHub action will update everything automatically
> 
> what do you think?

`brew` supports the `--no-quarantine` option, which seems to be a potential workaround for unsigned software. However, I haven't tested it myself, so I'm not sure if it will work correctly. If @love-linger is willing and agrees to publish SourceGit on Homebrew Cask, I could look into how to write a brew [Ruby script](https://docs.brew.sh/Cask-Cookbook) and submit a PR to [homebrew-cask](https://github.com/Homebrew/homebrew-cask).

Of course, it would be even better if someone with experience here could assist! :rocket:

**Comment by love-linger** [comment]
> > Maybe just pushing the release to some tools will do the job Each OS has its tools We don’t need more than that
> > for MacOS I can help pushing release process to homebrew…. like using this nice GitHub action: https://github.com/marketplace/actions/homebrew-bump-cask requires just creating a simple repo - with predefined structure.. and some ruby script - for installation the GitHub action will update everything automatically
> > what do you think?
> 
> `brew` supports the `--no-quarantine` option, which seems to be a potential workaround for unsigned software. However, I haven't tested it myself, so I'm not sure if it will work correctly. If @love-linger is willing and agrees to publish SourceGit on Homebrew Cask, I could look into how to write a brew [Ruby script](https://docs.brew.sh/Cask-Cookbook) and submit a PR to [homebrew-cask](https://github.com/Homebrew/homebrew-cask).
> 
> Of course, it would be even better if someone with experience here could assist! 🚀

Thanks very much. I'm not quite familier with Ruby and homebrew-cask.

**Comment by aikawayataro** [comment]
> brew supports the --no-quarantine option, which seems to be a potential workaround for unsigned software. However, I haven't tested it myself, so I'm not sure if it will work correctly. If @love-linger is willing and agrees to publish SourceGit on Homebrew Cask, I could look into how to write a brew [Ruby script](https://docs.brew.sh/Cask-Cookbook) and submit a PR to [homebrew-cask](https://github.com/Homebrew/homebrew-cask).

https://docs.brew.sh/Acceptable-Casks
>Common reasons to reject a cask entirely:
 App fails with GateKeeper enabled on Homebrew supported macOS versions and platforms (e.g. unsigned apps fail on Macs with Apple Silicon/ARM).

And if you scroll for rejected pull requests in homebrew-cask, you will see that a signing is required. Sadly, the only thing you can do without signing is to host your own tap (say repository).

**Comment by love-linger** [comment]
When I try to register an Apple Developer account, it always failes with this
<img width="752" alt="截屏2024-09-02 17 59 55" src="https://github.com/user-attachments/assets/ed441af2-0785-4bf8-bded-ac76170cf0d7">
So for this issue, I basically gave up...

**Comment by ybeapps** [comment]
If I’ll find someone with account
What do I need to provide from him and how can we keep his private keys only private for him?

**Comment by ChiahongHong** [comment]
I believe it's not a good idea to use a non-author's certificate for signing. It could lead to complications if someone reports a violation of any Apple policies later on.

**Comment by ChiahongHong** [comment]
I believe it's not a good idea to use a non-author's certificate for signing. It could lead to complications if someone reports a violation of any Apple policies later on.

Before @love-linger enroll in the Apple Developer Program, as mentioned earlier, that we can use Homebrew Tap as a transitional plan. Additionally, the annual fee of $99 isn't exactly cheap; ~~if it were me, I wouldn't want to spend that money either~~.

If @love-linger is willing to maintain the Tap, you could create a new repo under the `sourcegit-scm` organization with the name `homebrew-sourcegit`. Please copy the contents of my repo into this new one (instead of forking it directly, as that might not appear as official). Also, make sure to replace the `ChiahongHong/sourcegit` in the `README.md` with `sourcegit-scm/sourcegit`

 https://github.com/ChiahongHong/homebrew-sourcegit

You may adjust the `depends_on macos: ">= :catalina"` value as appropriate. The Avalonia UI documentation specifies a minimum of Mojave, and I used Catalina as a placeholder without actual testing.

Currently, the issue is that every time a new version of SourceGit is released, `sourcegit.rb` needs to be manually updated. If anyone knows how to integrate GitHub Actions to automatically run `bump-cask-pr`, that would be great!

**Comment by ybeapps** [comment]
I’ll try to find the time and try

**Comment by ybeapps** [comment]
my part is done and 90% percent tested..
we need to wrap it up and verify that everything is ok

**Comment by love-linger** [comment]
> I believe it's not a good idea to use a non-author's certificate for signing. It could lead to complications if someone reports a violation of any Apple policies later on.

I don't know much about it. The `sourcegit-scm` account is also my own and registered for this project only.

I was really tired with this. I don't want to go any further. I think it's pretty good now. The users  are basically developers. If they like this tool, I do not think it is a big problem to install or update this app manually.

**Comment by ChiahongHong** [comment]
@love-linger 感覺您誤會我的意思，這不是在說您，改用中文解釋一下，我是在回覆

> If I’ll find someone with account. What do I need to provide from him and how can we keep his private keys only private for him?`

他的意思看起來是想以別人的憑證代簽，我認為不妥才會這樣回答。

Anyway, we all respect your decision. Perhaps, as you said, it's already good enough.

**Comment by love-linger** [comment]
Hi, everyone. I'm going to close this issue.

**Comment by bernd-wechner** [comment]
I'd rather you left it open with a "Help wanted" sign. I mean just because it's a little tricky and not a priority now doesn't make it a bad idea. Just saying ...

**Comment by ybeapps** [comment]
@bernd-wechner
@ChiahongHong
@aikawayataro

thanks to all your help - homebrew support is here :)
see readme on `develop` branch

or just run the following:
```
brew tap ybeapps/homebrew-sourcegit
brew install --cask --no-quarantine sourcegit
```
